Doc: Add a design doc for optimizing gc safepoint advance (#32934)
pingcap/tidb: docs/design/2022-03-09-optimize-gc-for-internal-transaction.md

# Optimize GC Advance For Internal Transactions

- Author(s): [Michael](https://github.com/TonsnakeLin) 
- Tracking Issue: https://github.com/pingcap/tidb/issues/32725

## Introduction

This document describes the design of the feature "optimize gc advance for internal transaction", which adds the startTS(start timestamp) of all internal transactions to the gc(garbage collect) safepoint calculation process.

## Motivation or Background

> TiDB advances gc safe point every `tidb_gc_run_interval` time with the value `now - tidb_gc_life_time`. If there is a long time transaction from user client lives more than `tidb_gc_life_time`, the safepoint can't be advanced until the transaction is finished or lives over 24 hours. This mechanism ensures the continuous advancement of gc safepoint and ensures that the data active transactions need to access will not be cleared.

Reviewer: ```suggestion
TiDB advances gc safepoint every `tidb_gc_run_interval` time with the value `now - tidb_gc_life_time`. If there is a long time transaction from user client lives more than `tidb_gc_life_time`, the safepoint can't be advanced until the transaction is finished or lives over 24 hours. This mechanism ensures the continuous advancement of gc safepoint and ensures that the data active transactions need to access will not be cleared.
```

There are still some "safe point" not unified. I have no idea which is better though, let's adopt @you06 's suggestion here.

## Detailed Design

`Design Summary` describes the main processes of safepoint calculation and what job this design did.

### Design Summary

Currently here are two kinds of internal transactions. one is run by internal session, the other is run by function `RunInNewTxn`. This design aims to take account of all internal transactions when calculate gc safepoint.

The design stores internal sessions to session manager and stores the internal transactions to `globalInnerTxnTsBox` so that it can get startTS of internal transactions when calculates gc safepoint.

#### Calculate GC SafePoint

Currently, the processes TiDB calculates safepoint is as bellow, it is implemented in `(is *InfoSyncer) ReportMinStartTS`

- Get current timestamp from PD, save it to variable `now`,	 get timestamp that is 24 hours earlier than `now`, save it to variable `startTSLowerLimit`
- Get all user client sessions, save them to slice `processlist`
- Traverse `processlist` to get startTS from every client session and compare the `startTS` with `now` to get the minimum timestamp which is after `startTSLowerLimit` , save it as minStartTS which is gc safepoint.

> This design add extra processes in `(is *InfoSyncer) ReportMinStartTS` to take account of internal transactions startTS when calculates gc safe point . The new processes TiDB calculates safe point is as bellow.

Reviewer: ```suggestion
This design add extra processes in `(is *InfoSyncer) ReportMinStartTS` to take account of internal transactions startTS when calculates gc safepoint. The new processes TiDB calculates safepoint is as bellow.
```

```go
func (s *session) getInternalSession(execOption sqlexec.ExecOption) (*session, func(), error) {
	tmp, err := s.sysSessionPool().Get()
	se := tmp.(*session)
	// Put the internal session to the map of SessionManager
	infosync.StoreInternalSession(se)
	return se, func() {
		// Delete the internal session to the map of SessionManager
		infosync.DeleteInternalSession(se)
		s.sysSessionPool().Put(tmp)
	}
}
```

#### Store And Delete Internal Transactions

The internal transactions here are executed by function `RunInNewTxn`. It stores the internal transaction startTS by function  `wrapStoreInterTxnTS()` and deletes the transaction startTS by function `wrapDeleteInterTxnTS()`.

```go
// RunInNewTxn will run the f in a new transaction environment.
func RunInNewTxn(ctx context.Context, store Storage, retryable bool, f func(ctx context.Context, txn Transaction) error) error {
	defer func() {
		wrapDeleteInterTxnTS(originalTxnTS)
	}()
	for i := uint(0); i < maxRetryCnt; i++ {
		txn, err = store.Begin()

> 				if i == 0 {

Reviewer: ```suggestion
		if i == 0 {
```

			originalTxnTS = txn.StartTS()
			wrapStoreInterTxnTS(originalTxnTS)
		}
		err = txn.Commit(ctx)
		...
	}
}
```

> #### Calculate GC Safe Point

Reviewer: ```suggestion
#### Calculate GC Safepoint
```

> `ReportMinStartTS` calculates the minimum startTS of ongoing transactions (except the ones that exceed the limit), and it's result is useful for calculating the safe point, which is gc_worker's work.This design add some code in the  function `ReportMinStartTS` to consider internal transactions when calculates the minimum startTS.

Reviewer: ```suggestion
`ReportMinStartTS` calculates the minimum startTS of ongoing transactions (except the ones that exceed the limit), and it's result is useful for calculating the safepoint, which is gc_worker's work.This design add some code in the  function `ReportMinStartTS` to consider internal transactions when calculates the minimum startTS.
```

# Optimize GC Advance For Internal Transactions

- Author(s): [Michael](https://github.com/TonsnakeLin) 
- Tracking Issue: https://github.com/pingcap/tidb/issues/32725

## Introduction

> This document describes the design of the feature "optimize gc advance for internal transaction", which adds the startTS(start time stamp) of all internal transactions to the gc(garbage collect) safe point calculation process.

Reviewer 2: I see some "time stamp" and "safe point" usage, and I prefer "timestamp" and "safepoint".

Author: done

## Motivation or Background

> TiDB advances gc safe point every `tidb_gc_run_interval` time with step `tidb_gc_life_time`. If there is a long time transaction from user client lives more than `tidb_gc_life_time`, the safe point can't be advanced until the transaction is finished or lives over 24 hours. This mechanism ensures the continuous advancement of gc safe point and ensures that the data active transactions need to access will not be cleared.

Author: done

> However, Internal transactions run in TiDB don't comply with the mechanism above. If the internal transaction lives more than `tidb_gc_life_time`, maybe failed because the data it needs to access was cleared. This design aims to resolve the problem that internal transaction run failed because data is cleared by the gc mechanism.

Author: Thanks for checking in great detail.

### Data Structure Design

#### SessionManager 

TiDB manages client sessions by SessionManager interface. This design expands the SessionManager interface to manage internal sessions too.

```go
type SessionManager interface {
	// Put the internal session pointer to the map in the SessionManager
	StoreInternalSession(se interface{})
	// Delete the internal session pointer from the map in the SessionManager
	DeleteInternalSession(se interface{})
	// Get all startTS of every transactions running in the current internal sessions
	GetInternalSessionStartTSList() []uint64
}
```

#### innerTxnStartTsBox 

This design defines a global variable `globalInnerTxnTsBox` to manage internal transactions run by function `RunInNewTxn`. The map `innerTxnStartTsBox.innerTxnStartTsMap` stores all the internal transaction startTS. 

```go
var globalInnerTxnTsBox = innerTxnStartTsBox{
	innerTSLock:        sync.Mutex{},
	innerTxnStartTsMap: make(map[uint64]struct{}, 256),
}

type innerTxnStartTsBox struct {
	innerTSLock        sync.Mutex
	innerTxnStartTsMap map[uint64]struct{}
}
```

### Functional design

#### Store And Delete Internal Sessions

> TiDB gets an internal session from a session pool implemented in `(s *session) getInternalSession`, when the transaction finished,it puts the session to the pool. This design stores the internal session to session manger and deletes the internal session from session manager in the function  `(s *session) getInternalSession`. It calls the function `infosync.DeleteInternalSession()` to delete the internal session from session manager and calls the function `infosync.StoreInternalSession` to add the internal session to session manger.

Author: done

```go
func (s *session) getInternalSession(execOption sqlexec.ExecOption) (*session, func(), error) {
	tmp, err := s.sysSessionPool().Get()
	se := tmp.(*session)
	// Put the internal session to the map of SessionManager
	infosync.StoreInternalSession(se)
	return se, func() {
		// Delete the internal session to the map of SessionManager
		infosync.DeleteInternalSession(se)
		s.sysSessionPool().Put(tmp)
	}
}
```

#### Store And Delete Internal Transactions

The internal transactions here are executed by function `RunInNewTxn`. It stores the internal transaction startTS by function  `wrapStoreInterTxnTS()` and deletes the transaction startTS by function `wrapDeleteInterTxnTS()`.

```go
// RunInNewTxn will run the f in a new transaction environment.
func RunInNewTxn(ctx context.Context, store Storage, retryable bool, f func(ctx context.Context, txn Transaction) error) error {
	defer func() {
		wrapDeleteInterTxnTS(originalTxnTS)
	}()
    for i := uint(0); i < maxRetryCnt; i++ {
        txn, err = store.Begin()
        		if i == 0 {
			originalTxnTS = txn.StartTS()
			wrapStoreInterTxnTS(originalTxnTS)
		}
		err = txn.Commit(ctx)
        ...

>     }

Reviewer 2: Mixed usage of tabs and spaces. Replace spaces with tabs.

Author: done

}
```

#### Calculate GC Safe Point

> Currently, TiDB calculates gc safe point in the function `(is *InfoSyncer) ReportMinStartTS`. This design add some code in the  function `ReportMinStartTS` to consider internal transactions when calculates gc safe point.

Reviewer: I'd prefer saying `ReportMinStartTS` calculates the minimum startTS of ongoing transactions (except the ones that exceed the limit), and it's result is useful for calculating the safe point, which is gc_worker's work. The `ReportMinStartTS` doesn't calculate the safe point directly.

	undeletedTsLock        sync.Mutex
	chanToProcUndelStartTS chan uint64
	undeletedStartTsMap    map[uint64]uint64
}
```

### Functional design

#### Initialize Internal Transaction Global Management Box

This design initializes the global variable `globalInnerTxnTsBox` and starts 3 goroutines in the `main()` function to manage internal transactions run by function `RunInNewTxn()`. The initialization is wrapped in the function `InitInnerTxnStartTsBox()`. It closes the relevant goroutines in the function `cleanup()`

```go
func main() {
	kv.InitInnerTxnStartTsBox()
	signal.SetupSignalHandler(func(graceful bool) {
        svr.Close()
        cleanup(svr, storage, dom, graceful)
        cpuprofile.StopCPUProfiler()
        close(exited)
    })
}
```

the main code for `InitInnerTxnStartTsBox()` is as follower. It starts `storeInnerTxnStartTsLoop` goroutine to store an internal transaction startTS in `globalInnerTxnTsBox.innerTxnStartTsMap`, starts `deleteInnerTxnStartTsLoop` goroutine to delete an internal transaction startTS from `globalInnerTxnTsBox.innerTxnStartTsMap`. In particular, It starts `processUndeletedStartTsLoop` to delete the startTS for some rarely scenes.

```go

> func InitInnerTxnStartTsBox() {

Reviewer 3: Using background worker will introduce extra work like:
- When to close or shutdown properly, for example the `context` processing.
- Necessity to consider golang panic revcover to make sure the work is alive.
- Setting the channel size properly.

Do we have tests such as allocating auto-increment value every time to verify the overhead indroduced by putting ts into a concurrent map? Another simple way to ease the contesion is to shard the locks too.

- Traverse `innerTxnStartTsMap` to get startTS and compare with `minStartTS` to get the minimum time stamp which is after `startTSLowerLimit` , save it also as `minStartTS` which is gc safe point.

#### store and delete internal sessions

TiDB gets an internal session from a session pool implemented in `(s *session) getInternalSession`, when the transaction finished,it puts the session to the pool. It stores the session to session manager when gets an internal session, delete the session from session manager when returns the session to a pool.

#### store and delete internal transactions

The internal transactions here are executed by function `RunInNewTxn`. It stores the transaction to  `globalInnerTxnTsBox`  at transaction begin and deletes the transaction at end.

### data structure design

#### SessionManager 

TiDB manages client sessions by SessionManager interface. This design expands the SessionManager interface to manage internal sessions too.

```go
type SessionManager interface {
	// Put the internal session pointer to the map in the SessionManager

> 	StoreInternalSession(addr unsafe.Pointer)

Author: If so, I think it should use `interface{}` or others. 
Does the problem " the object which is cast to unsafe.Pointer may be GC" exist? @cfzjywxk @youjiali1995 @sticnarf @MyonKeminta

> 	StoreInternalSession(addr unsafe.Pointer)

Reviewer 4: It depends on how you implement the method. If the `unsafe.Pointer` still exists, it can be recognized by the GC.
The code in https://github.com/pingcap/tidb/pull/32726 looks good because the pointer is still held by the map.

But I don't really get why you need an `unsafe.Pointer` here... Isn't a simple `*session` sufficient?

> 	StoreInternalSession(addr unsafe.Pointer)

Author: if define `internalSessions map[*session]struct{} in `Server` struct, it will produce package cycle

> 	StoreInternalSession(addr unsafe.Pointer)

Reviewer 3: How about using the similar definition like the session pool.
The `usafe.Pointer` like parameters would make the interface difficult to understand and use.

	// Get all startTS of every transactions running in the current internal sessions
	GetInternalSessionStartTSList() []uint64
}
```

#### innerTxnStartTsBox 

This design defines a global variable `globalInnerTxnTsBox` to manage internal transactions run by function `RunInNewTxn`. The map `innerTxnStartTsBox.innerTxnStartTsMap` stores all the internal transaction startTS. 

```go
var globalInnerTxnTsBox atomic.Value

type innerTxnStartTsBox struct {
	wg                  sync.WaitGroup
	chanToStoreStartTS  chan uint64
	chanToDeleteStartTS chan uint64

	innerTxnTsBoxRun atomic.Value
	
	innerTSLock        sync.Mutex
	innerTxnStartTsMap map[uint64]uint64
	
	// `storeInnerTxnStartTsLoop` and `deleteInnerTxnStartTsLoop` run asynchronously,It can't ensure  
    //`storeInnerTxnStartTsLoop` receives startTS before `deleteInnerTxnStartTsLoop`,though 
    // `RunInNewTxn`send startTS to `storeInnerTxnStartTsLoop` firstly. If `deleteInnerTxnStartTsLoop`
	// recevied startTS before `storeInnerTxnStartTsLoop`, the `startTS` couldn't be deleted from 
    // `innerTxnStartTsMap`,and GC safepoint can't be advanced properly.
	undeletedTsLock        sync.Mutex
	chanToProcUndelStartTS chan uint64

> 	undeletedStartTsMap    map[uint64]uint64

Author: What's the benefits？ save memory?

> 	undeletedStartTsMap    map[uint64]uint64

Reviewer 2: Here what we want is less than a "map", actually a "set", and I think `map[uint64]struct{}` can be understood easily as a "set".

> 	undeletedStartTsMap    map[uint64]uint64

Reviewer 3: It's a commonly used format to represent `set` in golang.

}
```

### Functional design

#### Initialize internal transaction global management box

This design initializes the global variable `globalInnerTxnTsBox` and starts 3 goroutines in the `main()` function to manage internal transactions run by function `RunInNewTxn()`. The initialization is wrapped in the function `InitInnerTxnStartTsBox()`. It closes the relevant goroutines in the function `cleanup()`

```go
func main() {
	kv.InitInnerTxnStartTsBox()
    signal.SetupSignalHandler(func(graceful bool) {
        svr.Close()
        cleanup(svr, storage, dom, graceful)
        cpuprofile.StopCPUProfiler()
        close(exited)
    })
}
```

> the main code for `InitInnerTxnStartTsBox()` is as follower. It starts `storeInnerTxnStartTsLoop` goroutine to store an internal transaction startTS in `globalInnerTxnTsBox.innerTxnStartTsMap`, starts `deleteInnerTxnStartTsLoop` goroutine to delete an internal transaction startTS from `globalInnerTxnTsBox.innerTxnStartTsMap`. In particular, It starts `processUndeletedStartTsLoop` to delete the startTS for some rarely scenes.

Author: The two goroutines `storeInnerTxnStartTsLoop` and `deleteInnerTxnStartTsLoop` runs asynchronously. Though the `InnerTxnStartTs` is sent to `storeInnerTxnStartTsLoop` firstly and then sent to `deleteInnerTxnStartTsLoop` , But it can't ensure which goroutine received `InnerTxnStartTs`  firstly. If `deleteInnerTxnStartTsLoop` received `InnerTxnStartTs`  first, `InnerTxnStartTs`  cant be deleted forever

```go
func (s *session) getInternalSession(execOption sqlexec.ExecOption) (*session, func(), error) {
	tmp, err := s.sysSessionPool().Get()
	se := tmp.(*session)
	// Put the internal session to the map of SessionManager
	infosync.StoreInternalSession(unsafe.Pointer(se))
	return se, func() {
		// Delete the internal session to the map of SessionManager
		infosync.DeleteInternalSession(unsafe.Pointer(se))
		s.sysSessionPool().Put(tmp)
    }
}
```

#### store and delete internal transactions

The internal transactions here are executed by function `RunInNewTxn`. It stores the internal transaction startTS by function  `wrapStoreInterTxnTS()` and deletes the transaction startTS by function `wrapDeleteInterTxnTS()`.

```go
// RunInNewTxn will run the f in a new transaction environment.
func RunInNewTxn(ctx context.Context, store Storage, retryable bool, f func(ctx context.Context, txn Transaction) error) error {
	defer func() {
		wrapDeleteInterTxnTS(originalTxnTS)
	}()
    for i := uint(0); i < maxRetryCnt; i++ {
        txn, err = store.Begin()
        		if i == 0 {
			originalTxnTS = txn.StartTS()
			wrapStoreInterTxnTS(originalTxnTS)
		}
		err = txn.Commit(ctx)

>         ..........................................

Reviewer 2: Replace indent space with tab, and IMO, 3 dots are enough to represent omitted codes.

Author: Got it

    }
}
```

#### calculate gc safe point

Currently, TiDB calculates gc safe point in the function `(is *InfoSyncer) ReportMinStartTS`. This design add some code in the 

> 

Reviewer 2: Remove this empty line.

function `ReportMinStartTS` to consider internal transactions when calculates gc safe point.

```go

> func (is *InfoSyncer) ReportMinStartTS(store kv.Storage) {
> }

Reviewer 2: ```suggestion
func (is *InfoSyncer) ReportMinStartTS(store kv.Storage) {}
```
